### PR TITLE
Refactor handlers

### LINKED ISSUE #99: getRandSketchingMethod should be called setRandSketchingMethod
I think this should return the sketching method, not take it as input, no? I.e. this function should be called `set`

_Originally posted by @kswirydo in https://github.com/ORNL/ReSolve/pull/86#discussion_r1414762112_

### PR DESCRIPTION
**Pull request by pelesh**
This PR features refactored handler classes. The changes are as follows:
- Memory space of the handler is set upon instantiation.
- Host handler accompanies each device handler.
- Selection between device and host handler is done in the same way as for matrix/vector classes.
- All examples updated.
- All tests updated and passing.
- Closes #99.

### CODE REVIEW

File: resolve/vector/VectorHandlerCpu.cpp [hunk 1]
```diff
@@ -131,7 +131,7 @@ namespace ReSolve {
    * @pre   V is stored colum-wise, _n_ > 0, _k_ > 0
    * 
    */  
-  void VectorHandlerCpu::gemv(std::string /* transpose */,
+  void VectorHandlerCpu::gemv(char /* transpose */,
```
Comment (kswirydo) [inline resolve/vector/VectorHandlerCpu.cpp:134]: Why?

File: resolve/vector/VectorHandlerCpu.hpp [hunk 2]
```diff
@@ -45,7 +44,7 @@ namespace ReSolve { //namespace vector {
        * if `transpose = T` (yes), `x = beta*x + alpha*V^T*y`,
        * where `x` is `[k x 1]`, `V` is `[n x k]` and `y` is `[n x 1]`.
        */ 
-      virtual void gemv(std::string transpose,
+      virtual void gemv(char transpose,
```
Comment (kswirydo) [inline resolve/vector/VectorHandlerCpu.hpp:47]: Same here

File: resolve/vector/VectorHandlerCuda.cpp [hunk 3]
```diff
@@ -199,8 +198,7 @@ namespace ReSolve {
     if (k < 200) {
```
Comment (kswirydo) [inline resolve/vector/VectorHandlerCuda.cpp:198]: this constant would need to become a parameter at some point (but this is a future issue)

File: resolve/vector/VectorHandlerCuda.cpp [hunk 4]
```diff
@@ -239,8 +237,7 @@ namespace ReSolve {
...
```
Comment (kswirydo) [inline resolve/vector/VectorHandlerCuda.cpp:237]: and same here.

File: resolve/vector/VectorHandlerCuda.hpp [hunk 5]
```diff
@@ -45,7 +45,7 @@ namespace ReSolve { //namespace vector {
        * if `transpose = T` (yes), `x = beta*x + alpha*V^T*y`,
        * where `x` is `[k x 1]`, `V` is `[n x k]` and `y` is `[n x 1]`.
        */ 
-      virtual void gemv(std::string transpose,
+      virtual void gemv(char transpose,
```
Comment (kswirydo) [inline resolve/vector/VectorHandlerCuda.hpp:48]: why? I mean you should not compare chars like we compare strings,with `==`!!!

File: tests/unit/matrix/MatrixHandlerTests.hpp [hunk 6]
```diff
@@ -34,12 +34,17 @@ class MatrixHandlerTests : TestBase
   TestOutcome matrixInfNorm(index_type N)
   {
     TestStatus status;
+    ReSolve::memory::MemorySpace ms;
```
Comment (kswirydo) [inline tests/unit/matrix/MatrixHandlerTests.hpp:37]: what is the difference between `ms` and `memspace`??

File: tests/unit/vector/VectorHandlerTests.hpp [hunk 7]
```diff
@@ -92,7 +92,7 @@ namespace ReSolve {
 
           real_type alpha = 0.5;
           //the result is a vector with y[i] = 2.5;          
-          handler->axpy(&alpha, x, y, memspace_);
+          handler->axpy(&alpha, x, y, ms);
```
Comment (kswirydo) [inline tests/unit/vector/VectorHandlerTests.hpp:95]: again, what is the function of `ms` and `memspace`??

**Review by kswirydo — APPROVED** [review]

File: resolve/vector/VectorHandlerCpu.cpp (hunk 1 above)
Comment (pelesh) [reply resolve/vector/VectorHandlerCpu.cpp:134]: It is cheaper to compare `char` than `std::string`. Plus you can use `char` in `switch` statements.

File: tests/unit/matrix/MatrixHandlerTests.hpp (hunk 6 above)
Comment (pelesh) [reply tests/unit/matrix/MatrixHandlerTests.hpp:37]: `memspace` is string and `ms` is `enum`. I will eventually get rid of all strings.

File: tests/unit/vector/VectorHandlerTests.hpp (hunk 7 above)
Comment (pelesh) [reply tests/unit/vector/VectorHandlerTests.hpp:95]: Ditto.

File: resolve/vector/VectorHandlerCuda.cpp (hunk 3 above)
Comment (pelesh) [reply resolve/vector/VectorHandlerCuda.cpp:198]: See #112.

File: resolve/vector/VectorHandlerCuda.cpp (hunk 4 above)
Comment (pelesh) [reply resolve/vector/VectorHandlerCuda.cpp:237]: To be tracked in #112.

**Pull request merged**